Firmware/mmu2_protocol_logic.cpp

```diff
@@ -133,54 +125,92 @@ void ProtocolLogic::SendMsg(RequestMsg rq) {
 }
 
 void StartSeq::Restart() {
-    state = State::S0Sent;
-    logic->SendMsg(RequestMsg(RequestMsgCodes::Version, 0));
+    SendVersion(0);
 }
 
 StepStatus StartSeq::Step() {
-    auto expmsg = logic->ExpectingMessage(linkLayerTimeout);
-    if (expmsg != MessageReady)
+    if (auto expmsg = logic->ExpectingMessage(linkLayerTimeout); expmsg != MessageReady)
         return expmsg;
 
     // solve initial handshake
     switch (state) {
     case State::S0Sent: // received response to S0 - major
-        logic->mmuFwVersionMajor = logic->rsp.paramValue;
-        if (logic->mmuFwVersionMajor != 2) {
-            return VersionMismatch;
+        if( logic->rsp.request.code != RequestMsgCodes::Version || logic->rsp.request.value != 0 ){
+            // got a response to something else - protocol corruption probably, repeat the query
+                SendVersion(0);
```

Comment:
Does this repeated re-sending of versions have a chance of "stuffing" the event queue on the MMU and getting things out of sync? What we've seen in logs is usually a random other message followed by the correct S0 reply - would this not queue up a second S0 reply that might interfere with our expected S1 to follow?

Reply:
Good point, a few comments on that:
- That's one of the reasons we have the delayed start if the protocol goes out of sync.
- In case we "just" received a response to something else, repeating the query doesn't "stuff" the queue, because we always wait for the response from the MMU.
- If the MMU managed to join 2 responses together, the protocol may report an error at some spot.
But yes, in some rare cases the MMU might get overwhelmed by requests and will answer to each of them, potentially joining them into one sequence of response bytes.

Reply:
Excellent, if we clear the buffer and wait, then it is indeed not a problem. (unless there are bigger issues and the MMU is somehow hung up and takes a while to send its junk.